src/clj_irods/core.clj

```diff
@@ -137,6 +137,17 @@
       (let [jargon-stat (jargon/stat irods path)]
         (instrumented-delay (extract-fn @jargon-stat))))))
 
+(declare object-type)
+(defn- from-jargon-metadata
+  [cache? irods extract-fn post-fn user zone path]
+  (let [get-from-metadata (fn [metadata]
+                            (when-let [extracted (extract-fn @metadata)]
+                              (instrumented-delay (post-fn extracted))))]
+    (when-let [metadata (if cache?
+                          (jargon/cached-get-metadata irods path)
+                          (jargon/get-metadata irods path :known-type @(object-type irods user zone path)))]
+      (or (get-from-metadata metadata) nil))))
```

Comment:
No icat implementation of this yet. Right now, the main things that use metadata are uuid and info-type, which both can come from listings, so I haven't made it happen yet.